lib/wiki_assignment_output.rb

```diff
@@ -0,0 +1,97 @@
+    # only post to talk pages of mainspace articles, as we assume that pages
+    # like Talk:Template:Citation or Talk:Wikipedia:Notability do not exist.
+    if initial_page_content.nil?
+      return nil if Wiki.get_page_content(title).nil?
+      initial_page_content = ''
+    end
+
+    # Limit it to live assignments for this article-course
+    title_assignments = title_assignments.select { |a| !a['deleted'] }
+
+    course_assignments_tag = assignments_tag(course_page, title_assignments)
+    page_content = build_assignment_page_content(course_assignments_tag,
+                                                 course_page,
+                                                 initial_page_content)
+    page_content
+  end
+
+  ###################
+  # Helper methods #
+  ###################
+  def self.assignments_tag(course_page, title_assignments)
+    return '' if title_assignments.empty?
+
+    # Make a list of the assignees, role 0
+    tag_assigned = build_wikitext_user_list(title_assignments, 0)
+    # Make a list of the reviwers, role 1
+    tag_reviewing = build_wikitext_user_list(title_assignments, 1)
+
+    # Build new tag
+    # NOTE: If the format of this tag gets changed, then the dashboard may
+    # post duplicate tags for the same page, unless we update the way that
+    # we check for the presense of existging tags to account for both the new
+    # and old formats.
+    dashboard_url = Figaro.env.dashboard_url
+    tag = "{{#{dashboard_url} assignment | course = #{course_page}"
+    tag += " | assignments = #{tag_assigned}" unless tag_assigned.blank?
+    tag += " | reviewers = #{tag_reviewing}" unless tag_reviewing.blank?
+    tag += ' }}'
+
+    tag
+  end
+
+  # This method creates updated wikitext for an article talk page, for when
+  # the set of assigned users for the article for a single course changes.
+  # The strategy here is to only update the tag for one course at a time, so
+  # that the user who updates the assignments for a course only introduces data
+  # for that course. We also want to make as minimal a change as possible, and
+  # to make sure that we're not disrupting the format of existing content.
+  def self.build_assignment_page_content(new_tag,
+                                         course_page,
+                                         page_content)
+
+    dashboard_url = Figaro.env.dashboard_url
+    # Return if tag already exists on page
+    unless new_tag.blank?
+      return nil if page_content.include? new_tag
+    end
+
+    # Check for existing tags and replace
+    old_tag_ex = "{{course assignment | course = #{course_page}"
+    new_tag_ex = "{{#{dashboard_url} assignment | course = #{course_page}"
+    page_content.gsub!(/#{Regexp.quote(old_tag_ex)}[^\}]*\}\}/, new_tag)
+    page_content.gsub!(/#{Regexp.quote(new_tag_ex)}[^\}]*\}\}/, new_tag)
+
+    # Add new tag at top (if there wasn't an existing tag already)
+    if !page_content.include?(new_tag) && !new_tag.blank?
```

Comment:
That was before the gsub, in which we might have replaced an old tag with the new tag. This block is only for the case where we didn't already find somewhere else to put the new tag, so we want to add it to the top section of the page.